### PR TITLE
Add a task for warming up postcodes cache

### PR DESCRIPTION
**Pull request by catalinailie**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Trello card: https://trello.com/c/iy6P2rvX/2875-warm-up-locations-api-cache-5

In order to get the postcodes from mapit added into locations-api database, these steps need to be followed:
1. Export the postcodes from mapit db into a csv file:
`COPY (SELECT postcode FROM mapit_postcode limit 3) TO '/tmp/postcodes.csv' DELIMITER ',';`
2. Download the csv from mapit to the local machine:
`gds govuk connect scp-pull -e integration aws/mapit:1 /tmp/postcodes.csv ./ `
3. Upload the csv to the locations_api instance:
`gds govuk connect scp-push -e integration aws/locations_api:1 postcodes.csv /tmp`
4. Run the rake task from locations_api using Jenkins and pass `populate_postcodes_table\[/tmp/test.csv\]` as RAKE_TASK

### CODE REVIEW

**Comment by catalinailie** [comment]
This is the way I thought about this process:
1. Export `postcode` and `location` from `Postcode` table from mapit db in a csv file `mapi.csv`
2. Run the rake task which creates a new csv file `postcodes.csv`
3. Import `postcodes.csv` in `postcodes` table from locations-api

I have doubts regarding the time this process will take.
Also I saw that mapit already has a `.sql.gz` db export on S3, but I don't know how we can use this in our rake task.

File: lib/tasks/populate_postcodes_table.rake
Comment (ChrisBAshton) [inline lib/tasks/populate_postcodes_table.rake:10]: Fancy! Always enjoy a deconstruct operator 😁

File: lib/tasks/populate_postcodes_table.rake
Comment (ChrisBAshton) [inline lib/tasks/populate_postcodes_table.rake:5]: It would be nice to make this a user argument ([example](https://github.com/alphagov/whitehall/blob/126891bf9f2ff49877c5649aa069ca21575b23f6/lib/tasks/publishing_api.rake#L281-L291)), so that the dev can specify the absolute path to the CSV.

File: lib/tasks/populate_postcodes_table.rake
Comment (ChrisBAshton) [inline lib/tasks/populate_postcodes_table.rake:19]: Mapit uses a different algorithm for determining Lat/Long, IIRC. 🤔 

I think it would be easiest (and most consistent / safest) to just import the postcode, and not worry about anything else. I.e. `Postcode.create(postcode: postcode, results: "{}")`. Then we can let the `ProcessPostcodeWorker` do its work.

**Review by ChrisBAshton — COMMENTED** [review]
Good start, @catalinailie - I've left some pointers below.

Also I think the S3 file you're referring to is probably the database backup for Mapit. You can use it if you wish (might be kind of nice, as we then have a reliable data source we can point to and write docs on how to export it and manipulate it into the format we want). But it's also fine to extract the postcodes from Mapit in whatever way makes sense to you.

File: lib/tasks/populate_postcodes_table.rake
Comment (catalinailie) [reply lib/tasks/populate_postcodes_table.rake:19]: Yes, makes sense since the api is not used yet and we don't need all the information. 👍

File: lib/tasks/populate_postcodes_table.rake
Comment (catalinailie) [reply lib/tasks/populate_postcodes_table.rake:26]: Yes, I agree. I just wanted to make sure this is the correct path before doing the finishing touches. Thanks 🙂
Also I tried to insert into the db directly from the task, but I'm not sure if this is possible or how should I import the table model in the task.

Comment (ChrisBAshton) [reply lib/tasks/populate_postcodes_table.rake:26]: Hmmm, [this line](https://github.com/alphagov/locations-api/blob/main/Rakefile#L6) should handle all the class loading for us, so `Postcode.create(postcode: postcode, results: "{}")` should work 🤞

Comment (catalinailie) [reply lib/tasks/populate_postcodes_table.rake:26]: Yes, it works. I think last time I tried and failed it was because the db was not set up properly. Thanks

File: spec/lib/tasks/populate_postcodes_spec.rb
Comment (ChrisBAshton) [inline spec/lib/tasks/populate_postcodes_spec.rb:31]: I'd add one more test: that it doesn't overwrite any existing postcode rows.

e.g. `Postcode.create(postcode: "SE1 2TL", results: '{ "foo": "bar" }')` and assert that `the_postcode.results` is still the same after running the task.

File: lib/tasks/populate_postcodes_table.rake
Comment (ChrisBAshton) [inline lib/tasks/populate_postcodes_table.rake:5]: I'd call the argument `path_to_csv` or something 👍  Otherwise looks like we're passing an array of postcodes.

**Review by ChrisBAshton — COMMENTED** [review]
This is coming along nicely ⭐ Yes, don't worry about the file in the S3 bucket - sounds like it would be too much effort to unzip the file and extract the CSVs from the SQL file. I would document how you can extract postcodes from Mapit though - and what you need to do to convert it to just the CSV of postcodes.

File: spec/lib/tasks/populate_postcodes_spec.rb
Comment (ChrisBAshton) [inline spec/lib/tasks/populate_postcodes_spec.rb:40]: This can be tidied up to work with just one element instead of an array of results:

```suggestion
      expect(Postcode.find_by(postcode: "SE12TL").results).to eq(existing_result)
```

File: spec/lib/tasks/populate_postcodes_spec.rb
Comment (ChrisBAshton) [inline spec/lib/tasks/populate_postcodes_spec.rb:1]: Nitpick: I'd rename this file to be consistent with the rake task file, i.e. `populate_postcodes_table_spec.rb`.

File: spec/lib/tasks/populate_postcodes_spec.rb
Comment (ChrisBAshton) [inline spec/lib/tasks/populate_postcodes_spec.rb:34]: I'd cache the existing result in a variable so that we can assert against that variable later:

```suggestion
      existing_result = '{ "foo": "bar" }'
      Postcode.create(postcode: "SE12TL", results: existing_result)
```

**Review by ChrisBAshton — COMMENTED** [review]
A couple more final tweaks before I'll approve and merge ✅ 

This should now be ready to test on Integration too (now the workers are sorted out again).

**Review by ChrisBAshton — APPROVED** [review]
Approved, pending a test on integration ⭐

**Pull request merged**
